File: app/task_calendar.py
from __future__ import annotations

from datetime import date

STATUSES = frozenset({"pending", "completed", "skipped"})
REMINDER_STATES = frozenset({"overdue", "due_today", "upcoming"})


def _iso_date(value: str, label: str) -> date:
    text = str(value or "").strip()
    try:
        parsed = date.fromisoformat(text)
    except ValueError as exc:
        raise ValueError(f"{label} must be YYYY-MM-DD") from exc
    if parsed.isoformat() != text:
        raise ValueError(f"{label} must be YYYY-MM-DD")
    return parsed


def task_state(
    status: str,
    due_date: str,
    today: date | str,
    *,
    horizon_days: int = 7,
) -> str:
    """Return the Phase 13 effective task state without mutating stored status."""
    value = str(status or "").strip()
    if value not in STATUSES:
        raise ValueError(f"Unsupported crop task status: {value}")
    if horizon_days < 0:
        raise ValueError("horizon_days must be nonnegative")

    due = _iso_date(due_date, "due_date")
    current = _iso_date(today, "today") if isinstance(today, str) else today
    if not isinstance(current, date):
        raise ValueError("today must be a date or YYYY-MM-DD")

    if value in {"completed", "skipped"}:
        return value
    if due < current:
        return "overdue"
    if due == current:
        return "due_today"
    if (due - current).days <= horizon_days:
        return "upcoming"
    return "pending"


def reminder_severity(
    status: str,
    due_date: str,
    today: date | str,
    *,
    horizon_days: int = 7,
) -> int | None:
    """0=urgent overdue, 1=due today, 2=upcoming, None=no reminder."""
    state = task_state(status, due_date, today, horizon_days=horizon_days)
    return {"overdue": 0, "due_today": 1, "upcoming": 2}.get(state)

File: android/app/src/main/java/gr/mastixa/manager/TaskCalendar.java
package gr.mastixa.manager;

import java.time.LocalDate;
import java.time.format.DateTimeParseException;
import java.util.Set;

/** Phase 13 derived calendar/reminder semantics. Stored task status is never rewritten to overdue. */
public final class TaskCalendar {
    public static final Set<String> STATUSES = Set.of("pending", "completed", "skipped");
    public static final Set<String> REMINDER_STATES = Set.of("overdue", "due_today", "upcoming");

    private TaskCalendar() {}

    private static LocalDate isoDate(String value, String label) {
        String text = value == null ? "" : value.trim();
        try {
            LocalDate parsed = LocalDate.parse(text);
            if (!parsed.toString().equals(text)) throw new DateTimeParseException("non-canonical", text, 0);
            return parsed;
        } catch (DateTimeParseException error) {
            throw new IllegalArgumentException(label + " must be YYYY-MM-DD", error);
        }
    }

    public static String state(String status, String dueDate, LocalDate today, int horizonDays) {
        String value = status == null ? "" : status.trim();
        if (!STATUSES.contains(value))
            throw new IllegalArgumentException("Unsupported crop task status: " + value);
        if (today == null) throw new IllegalArgumentException("today is required");
        if (horizonDays < 0) throw new IllegalArgumentException("horizon_days must be nonnegative");

        LocalDate due = isoDate(dueDate, "due_date");
        if (value.equals("completed") || value.equals("skipped")) return value;
        if (due.isBefore(today)) return "overdue";
        if (due.equals(today)) return "due_today";
        if (!due.isAfter(today.plusDays(horizonDays))) return "upcoming";
        return "pending";
    }

    public static String state(String status, String dueDate, LocalDate today) {
        return state(status, dueDate, today, 7);
    }

    public static Integer reminderSeverity(String status, String dueDate, LocalDate today, int horizonDays) {
        return switch (state(status, dueDate, today, horizonDays)) {
            case "overdue" -> 0;
            case "due_today" -> 1;
            case "upcoming" -> 2;
            default -> null;
        };
    }

    public static Integer reminderSeverity(String status, String dueDate, LocalDate today) {
        return reminderSeverity(status, dueDate, today, 7);
    }
}
